bob-access/src/authenticator/basic.rs

```diff
@@ -1,65 +1,136 @@
+    }
+
+    async fn resolve_worker(nodes: Arc<RwLock<NodesCredentials>>, cred: DeclaredCredentials, sleep_dur_ms: u64) {
+        let hostname = cred.hostname().as_ref().expect("resolve worker without hostname");
+        let mut addr: Option<Vec<SocketAddr>> = None;
+        let mut cur_sleep_dur_ms = 100;
+        while addr.is_none() || (addr.is_some() && addr.as_ref().unwrap().len() == 0) {
+            tokio::time::sleep(Duration::from_millis(cur_sleep_dur_ms)).await;
+
+            addr = match lookup_host(hostname).await {
+                Ok(address) => Some(address.collect()),
+                _ => None
+            };
+
+            cur_sleep_dur_ms = sleep_dur_ms.min(cur_sleep_dur_ms * 2);
         }
-        self.nodes
-            .get(ip.as_ref()?)
-            .map(|creds| {
-                for cred in creds {
-                    if let Some(CredentialsKind::InterNode(other_name)) = cred.kind() {
-                        if node_name == other_name {
-                            return true;
-                        }
+
+        let addr = addr.expect("somehow addr is none");
+        let mut nodes = nodes.write().expect("nodes credentials lock");
+        if let Some(CredentialsKind::InterNode(nodename)) = cred.kind() {
+            if let Some(creds) = nodes.get_mut(nodename) {
+                creds.add_addresses(addr);
+                creds.set_resolved();
+            }
+        } else {
+            error!("resolved credentials are not internode");
+        }
+    }
+
+    fn check_resolve(&self, nodename: &str, unresolved: bool) {
+        let mut nodes = self.nodes.write().expect("nodes credentials lock");
+        if let Some(node) = nodes.get_mut(nodename) {
+            if node.needs_resolve(unresolved) {
+                node.set_pending();
+                self.spawn_resolver(node.clone());
+            }
+        }
+    }
+
+    fn check_node_request(&self, node_name: &String, ip: Option<SocketAddr>) -> bool {
+        let mut unresolved = false;
+        {
+            if ip.is_none() {
+                return false;
+            }
+            let ip = ip.unwrap().ip();
+            let nodes = self.nodes.read().expect("nodes credentials lock");
+            if nodes.is_empty() {
+                warn!("nodes credentials not set");
+                return false;
+            }
+            if let Some(cred) = nodes.get(node_name) {
+                if let Some(CredentialsKind::InterNode(other_name)) = cred.kind() {
+                    debug_assert!(node_name == other_name);
+                    if cred.ip().as_ref().and_then(|ips| {
+                        ips.iter().find(|cred_ip| cred_ip.ip() == ip)
+                    }).is_some() {
+                        return true;
                     }
+                    
+                    unresolved = true;
                 }
-                false
-            })
+            }
+        }
+        self.check_resolve(node_name, unresolved);
```

Comment:
If there is no node with the specified name in `nodes`, you do not need to call check_resolve`